dialogue-annotations/src/test/java/com/palantir/dialogue/annotations/MultipartRequestBodyTest.java

```diff
@@ -0,0 +1,333 @@
+
+        MultipartBody.Builder multipartBodyBuilder = new MultipartBody.Builder(BOUNDARY);
+        multipartBodyBuilder.setType(okhttp3.MediaType.parse(javax.ws.rs.core.MediaType.MULTIPART_FORM_DATA));
+        RequestBody body = unknownLengthRequestBody(httpMediaContent, okhttp3.MediaType.parse(httpMediaType));
+        multipartBodyBuilder.addPart(MultipartBody.Part.createFormData(uploadField, httpMediaFileName, body));
+        RequestBody detailsBody = unknownLengthRequestBody(detailsBytes, okhttp3.MediaType.parse(detailsContentType));
+        multipartBodyBuilder.addPart(MultipartBody.Part.createFormData(detailsField, null, detailsBody));
+        MultipartBody okhttp = multipartBodyBuilder.build();
+
+        MultipartRequestBody dialogue = MultipartRequestBody.builder()
+                .boundary(BOUNDARY)
+                .addFormBodyPart(MultipartRequestBody.formBodyPartBuilder(
+                                uploadField, byteArrayUnknownLengthRequestBody(httpMediaType, httpMediaContent))
+                        .fileName(httpMediaFileName))
+                .addFormBodyPart(MultipartRequestBody.formBodyPartBuilder(
+                        detailsField, byteArrayUnknownLengthRequestBody(detailsContentType, detailsBytes)))
+                .build();
+
+        assertOkhttpAndDialogueMatch(okhttp, dialogue);
+    }
+
+    private com.palantir.dialogue.annotations.ContentBody byteArrayUnknownLengthRequestBody(
+            String contentType, byte[] value) {
+        return new com.palantir.dialogue.annotations.ContentBody() {
+            @Override
+            public void writeTo(OutputStream output) throws IOException {
+                output.write(value);
+            }
+
+            @Override
+            public String contentType() {
+                return contentType;
+            }
+
+            @Override
+            public void close() {}
+        };
+    }
+
+    private com.palantir.dialogue.annotations.ContentBody byteArrayLengthRequestBody(String contentType, byte[] value) {
+        return new com.palantir.dialogue.annotations.ContentBody() {
+            @Override
+            public void writeTo(OutputStream output) throws IOException {
+                output.write(value);
+            }
+
+            @Override
+            public String contentType() {
+                return contentType;
+            }
+
+            @Override
+            public OptionalLong contentLength() {
+                return OptionalLong.of(value.length);
+            }
+
+            @Override
+            public void close() {}
+        };
+    }
+
+    @Value.Immutable
+    @Value.Style(stagedBuilder = true)
+    interface KeyValue {
+        String key();
+
+        String bucket();
+
+        String value();
+
+        Map<String, String> keyValues();
+
+        String contentType();
+    }
+
+    private RequestBody unknownLengthRequestBody(byte[] value, okhttp3.MediaType contentType) {
```

Comment:
So all this ceremony is because apache does not set Content-Length header for each individual part. as discussed with @carterkozak it doesn't feel like a problem, so we'll have it on our ContentBody interface, in case it's supported in the future, but unlikely to be a big deal for anyone that has to use this.